docs/books/valhalla-mjolnir/src/03_graph_construction_process.md

```diff
@@ -0,0 +1,310 @@
+};
+```
+
+The enhancement process includes:
+
+1. **Adding Turn Lanes**: Information about lane configuration at intersections
+2. **Speed Assignment**: Setting speed limits based on road class and other factors
+3. **Access Restrictions**: Determining which modes of travel can use each edge
+4. **Adding Administrative Information**: Country, state, and other boundaries
+5. **Edge Classification**: Identifying ramps, links, internal edges, etc.
+
+Here's a simplified example of the enhancement process:
+
+```cpp
+// From src/mjolnir/graphenhancer.cc
+void GraphEnhancer::Enhance(...) {
+  // For each tile in the graph
+  for (const auto& tile_id : tile_ids) {
+    GraphTileBuilder tile_builder(tile_id);
+    
+    // For each edge in the tile
+    for (const auto& edge : tile_builder.directededges()) {
+      // Enhance with additional attributes
+      edge.set_speed(DetermineSpeed(edge));
+      edge.set_access(DetermineAccess(edge));
+      edge.set_classification(ClassifyEdge(edge));
+      
+      // Add turn lanes if available
+      if (HasTurnLanes(edge)) {
+        edge.set_turnlanes(GetTurnLanes(edge));
+      }
+      
+      // Update the edge in the builder
+      tile_builder.update_directededge(edge);
+    }
+    
+    // Save the enhanced tile
+    tile_builder.StoreTileData();
+  }
+}
+```
+
+The enhancement phase is where much of the routing intelligence is added to the graph. By analyzing the topology and attributes of the road network, Mjolnir can infer additional properties that aren't explicitly tagged in OSM data.
+
+## Complex Restriction Building
+
+Turn restrictions (e.g., no left turn, no U-turn) are crucial for accurate routing. Mjolnir processes these using the `ComplexRestrictionBuilder`:
+
+```cpp
+// From valhalla/mjolnir/complexrestrictionbuilder.h
+class ComplexRestrictionBuilder {
+public:
+  ComplexRestrictionBuilder();
+  
+  // Set various attributes
+  void set_from_id(const GraphId& id);
+  void set_to_id(const GraphId& id);
+  void set_via_id(const GraphId& id);
+  void set_type(const RestrictionType type);
+  // ...
+};
+```
+
+The implementation in `src/mjolnir/complexrestrictionbuilder.cc` handles the conversion from OSM restriction relations to Valhalla's internal representation. Restrictions can be:
+- Simple (from one edge to another)
+- Complex (involving multiple via edges or nodes)
+- Time-dependent (only active during certain hours)
+- Mode-specific (only applying to certain vehicle types)
+
+## Hierarchical Graph Building
+
+Valhalla's multi-level graph is created by the `HierarchyBuilder`:
+
+```cpp
+// From valhalla/mjolnir/hierarchybuilder.h
+class HierarchyBuilder {
+public:
+  static void Build(const boost::property_tree::ptree& pt);
+};
+```
+
+The process involves:
+
+1. **Edge Classification**: Determining which edges belong to which hierarchy level
+2. **Shortcut Creation**: Creating direct connections between important nodes to bypass less important roads
```

Comment:
this is also known as contraction, as is used in contraction hierarchies. its important to know that we do the most conservative thing we can which is to say we only contract nodes between edges that have no meaningful (for the sake of path computation) change in attribution

Reply:
also its a bit strange to mention shortcuts here. i would simply put them in their own section after the heirarchy building